src/py/flwr/superexec/exec_event_log_interceptor.py

```diff
@@ -0,0 +1,119 @@
+        # One of the method handlers in
+        # `flwr.superexec.exec_servicer.ExecServicer`
+        method_handler: grpc.RpcMethodHandler = continuation(handler_call_details)
+        method_name: str = handler_call_details.method
+        return self._generic_auth_unary_method_handler(method_handler, method_name)
+
+    def _generic_auth_unary_method_handler(
+        self, method_handler: grpc.RpcMethodHandler, method_name: str
+    ) -> grpc.RpcMethodHandler:
+        def _generic_method_handler(
+            request: EventLogRequest,
+            context: grpc.ServicerContext,
+        ) -> Union[EventLogResponse, Generator[Any, Any, None]]:
+            log_entry: LogEntry
+            # Log before call
+            log_entry = self.log_plugin.compose_log_before_event(
+                request,
+                context,
+                shared_user_info.get(),
+                method_name,
+            )
+            self.log_plugin.write_log(log_entry)
+
+            # For unary-unary calls, log after the call immediately.
+            if method_handler.unary_unary:
+                response = method_handler.unary_unary(request, context)
+                log_entry = self.log_plugin.compose_log_after_event(
+                    request,
+                    context,
+                    response,
+                    shared_user_info.get(),
+                    method_name,
+                )
+                self.log_plugin.write_log(log_entry)
+                return cast(EventLogResponse, response)
+
+            # For unary-stream calls, wrap the response iterator write the event log
+            # after iteration completes.
+            if method_handler.unary_stream:
+                response = cast(
+                    Iterator[EventLogResponse],
+                    method_handler.unary_stream(request, context),
+                )
+
+                def response_wrapper() -> Generator[Any, Any, None]:
+                    try:
+                        yield from response
+                    finally:
+                        # This block is executed after the client has consumed
+                        # the entire stream, or if iteration is interrupted.
+                        log_entry = self.log_plugin.compose_log_after_event(
+                            request,
+                            context,
+                            response,
+                            shared_user_info.get(),
+                            method_name,
+                        )
+                        self.log_plugin.write_log(log_entry)
+
+                return response_wrapper()
+
+            # If the method type is not `unary_unary` or `unary_stream`, raise an error.
+            raise NotImplementedError("This RPC method type is not supported.")
+
+        if method_handler.unary_unary:
+            message_handler = grpc.unary_unary_rpc_method_handler
+        else:
+            message_handler = grpc.unary_stream_rpc_method_handler
```

Comment:
Instead of raise an exception in the generated method handler, should we raise it here?
```suggestion
            # If the method type is not `unary_unary` or `unary_stream`, raise an error.
            raise NotImplementedError("This RPC method type is not supported.")

        if method_handler.unary_unary:
            message_handler = grpc.unary_unary_rpc_method_handler
        elif method_handler.unary_stream:
            message_handler = grpc.unary_stream_rpc_method_handler
        else:
            # If the method type is not `unary_unary` or `unary_stream`, raise an error.
            raise NotImplementedError("This RPC method type is not supported.")
```